File: ios/bitchat-mesh/Sources/BitchatMesh/MeshManager.swift
import CoreBluetooth
import Foundation

/// High-level mesh API aligned with the Android MeshManager surface.
public final class MeshManager {
    public weak var listener: MeshListener?

    private let keychainManager: KeychainManagerProtocol?
    private let identityManager: SecureIdentityStateManager?
    private let transport: Transport
    private var hasStarted: Bool = false
    private var transferProgressObserverId: UUID?

    public var myPeerID: PeerID {
        transport.myPeerID
    }

    public var myPeerId: String {
        myPeerID.id
    }

    public var myNickname: String {
        transport.myNickname
    }

    public init(configuration: MeshConfiguration = .default) {
        let keychain = KeychainManager(configuration: configuration)
        let identityManager = SecureIdentityStateManager(keychain)
        self.keychainManager = keychain
        self.identityManager = identityManager
        let service = BLEService(
            keychain: keychain,
            identityManager: identityManager
        )
        self.transport = service
        service.delegate = self
        VerificationService.shared.configure(with: service.getNoiseService())
        self.transferProgressObserverId = TransferProgressManager.shared.addObserver { [weak self] event in
            self?.listener?.onTransferProgress(
                transferId: event.transferId,
                sent: event.sent,
                total: event.total,
                completed: event.completed
            )
        }
    }

    init(transport: Transport) {
        self.keychainManager = nil
        self.identityManager = nil
        self.transport = transport
        transport.delegate = self
        VerificationService.shared.configure(with: transport.getNoiseService())
        self.transferProgressObserverId = TransferProgressManager.shared.addObserver { [weak self] event in
            self?.listener?.onTransferProgress(
                transferId: event.transferId,
                sent: event.sent,
                total: event.total,
                completed: event.completed
            )
        }
    }

    deinit {
        if let observerId = transferProgressObserverId {
            TransferProgressManager.shared.removeObserver(observerId)
        }
    }

    /// Assign a listener for mesh events.
    public func setListener(_ listener: MeshListener?) {
        self.listener = listener
    }

    /// Start BLE services and optionally set the local nickname.
    public func start(nickname: String? = nil) {
        if let nickname = nickname?.trimmingCharacters(in: .whitespacesAndNewlines), !nickname.isEmpty {
            transport.setNickname(nickname)
        }
        transport.startServices()
        hasStarted = true
        listener?.onStarted()
    }

    /// Stop BLE services.
    public func stop() {
        transport.stopServices()
        hasStarted = false
        listener?.onStopped()
    }

    /// Update the local nickname used in announces and messages.
    public func setNickname(_ nickname: String) {
        transport.setNickname(nickname)
    }

    /// Return true once services have started and not yet stopped.
    public func isRunning() -> Bool {
        hasStarted
    }

    /// Alias for `isRunning()` to match Android naming.
    public func isStarted() -> Bool {
        hasStarted
    }

    /// Send a broadcast chat message with optional mentions.
    /// - Note: `channel` is accepted for parity but is currently ignored on iOS.
    public func sendBroadcastMessage(_ content: String, mentions: [String] = [], channel: String? = nil) {
        _ = channel
        transport.sendMessage(content, mentions: mentions)
    }

    /// Send a private message to a peer.
    public func sendPrivateMessage(_ content: String, to peerID: PeerID, recipientNickname: String, messageID: String? = nil) {
        let id = messageID ?? UUID().uuidString
        transport.sendPrivateMessage(content, to: peerID, recipientNickname: recipientNickname, messageID: id)
        listener?.onSent(messageID: id, recipientPeerID: peerID.id)
    }

    /// Initiate the Noise handshake with a peer.
    public func establish(_ peerID: PeerID) {
        transport.triggerHandshake(with: peerID)
    }

    /// Alias for `establish(_:)`.
    public func triggerHandshake(with peerID: PeerID) {
        establish(peerID)
    }

    /// Return true if a Noise session is established with the peer.
    public func isEstablished(_ peerID: PeerID) -> Bool {
        transport.getNoiseService().hasEstablishedSession(with: peerID)
    }

    /// Send a file packet as a broadcast transfer.
    public func sendFileBroadcast(_ packet: BitchatFilePacket) {
        guard let payload = packet.encode() else { return }
        let transferId = payload.sha256Hex()
        transport.sendFileBroadcast(packet, transferId: transferId)
    }

    /// Send a file packet directly to a peer.
    public func sendFilePrivate(_ packet: BitchatFilePacket, to peerID: PeerID) {
        guard let payload = packet.encode() else { return }
        let transferId = payload.sha256Hex()
        transport.sendFilePrivate(packet, to: peerID, transferId: transferId)
    }

    public func cancelTransfer(_ transferId: String) {
        transport.cancelTransfer(transferId)
    }

    /// Send a read receipt to a peer.
    public func sendReadReceipt(_ receipt: ReadReceipt, to peerID: PeerID) {
        transport.sendReadReceipt(receipt, to: peerID)
    }

    /// Send a delivery acknowledgment to a peer.
    public func sendDeliveryAck(for messageID: String, to peerID: PeerID) {
        transport.sendDeliveryAck(for: messageID, to: peerID)
    }

    /// Current peer nickname map as known by the mesh.
    public func peerNicknames() -> [PeerID: String] {
        transport.getPeerNicknames()
    }

    /// Current peer RSSI map, if available.
    /// - Note: iOS does not expose RSSI yet, so this returns an empty map.
    public func peerRssi() -> [PeerID: Int] {
        [:]
    }
}

extension MeshManager: BitchatDelegate {
    func didReceiveMessage(_ message: BitchatMessage) {
        listener?.onMessageReceived(message)
        listener?.onReceived(message)
    }

    func didConnectToPeer(_ peerID: PeerID) {
        listener?.onFound(peerID)
        listener?.onPeerConnected(peerID)
    }

    func didDisconnectFromPeer(_ peerID: PeerID) {
        listener?.onLost(peerID)
        listener?.onPeerDisconnected(peerID)
    }

    func didUpdatePeerList(_ peers: [PeerID]) {
        listener?.onPeerListUpdated(peers)
    }

    func didUpdateMessageDeliveryStatus(_ messageID: String, status: DeliveryStatus) {
        switch status {
        case .delivered(let nickname, let at):
            listener?.onDeliveryAck(messageID: messageID, recipientNickname: nickname, timestamp: at)
        case .read(let nickname, let at):
            listener?.onReadReceipt(messageID: messageID, recipientNickname: nickname, timestamp: at)
        default:
            break
        }
    }

    func didReceiveNoisePayload(from peerID: PeerID, type: NoisePayloadType, payload: Data, timestamp: Date) {
        listener?.onNoisePayloadReceived(from: peerID, type: type, payload: payload, timestamp: timestamp)
        switch type {
        case .verifyChallenge:
            listener?.onVerifyChallenge(peerID: peerID, payload: payload, timestamp: timestamp)
        case .verifyResponse:
            listener?.onVerifyResponse(peerID: peerID, payload: payload, timestamp: timestamp)
        default:
            break
        }
    }

    func didUpdateBluetoothState(_ state: CBManagerState) {
        listener?.onBluetoothStateUpdated(state)
    }

    func didReceivePublicMessage(from peerID: PeerID, nickname: String, content: String, timestamp: Date, messageID: String?) {
        listener?.onPublicMessageReceived(from: peerID, nickname: nickname, content: content, timestamp: timestamp, messageID: messageID)
    }

    func didReceiveFileTransfer(peerID: PeerID, fileName: String, fileSize: Int, mimeType: String, localURL: URL) {
        listener?.onFileReceived(peerID: peerID, fileName: fileName, fileSize: fileSize, mimeType: mimeType, localURL: localURL)
    }
}

File: android/bitchat-mesh/src/main/java/com/bitchat/mesh/MeshManager.kt
package com.bitchat.mesh

import android.content.Context
import com.bitchat.android.mesh.BluetoothMeshDelegate
import com.bitchat.android.mesh.BluetoothMeshService
import com.bitchat.android.mesh.TransferProgressManager
import com.bitchat.android.model.BitchatFilePacket
import com.bitchat.android.model.BitchatMessage
import com.bitchat.android.services.NicknameProvider
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class MeshManager(private val context: Context) {
    private var service: BluetoothMeshService? = null
    private var listener: MeshListener? = null
    private var nickname: String? = null
    private var lastPeers: Set<String> = emptySet()
    private val progressScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    init {
        progressScope.launch {
            TransferProgressManager.events.collect { event ->
                listener?.onTransferProgress(event.transferId, event.sent, event.total, event.completed)
            }
        }
    }

    fun setListener(listener: MeshListener?) {
        this.listener = listener
        service?.delegate = buildDelegate()
    }

    fun start(nickname: String? = null) {
        this.nickname = nickname?.trim()?.ifEmpty { null }
        NicknameProvider.setNickname(this.nickname)
        val mesh = BluetoothMeshService(context.applicationContext)
        mesh.delegate = buildDelegate()
        mesh.startServices()
        service = mesh
        listener?.onStarted()
    }

    fun stop() {
        service?.stopServices()
        service = null
        lastPeers = emptySet()
        listener?.onStopped()
    }

    fun myPeerId(): String? = service?.myPeerID

    fun isRunning(): Boolean = service?.isReusable() == true

    fun isStarted(): Boolean = service != null && service?.isReusable() == true

    fun sendBroadcastMessage(content: String, mentions: List<String> = emptyList(), channel: String? = null) {
        service?.sendMessage(content, mentions, channel)
        listener?.onSent(null, null)
    }

    fun sendPrivateMessage(
        content: String,
        recipientPeerID: String,
        recipientNickname: String,
        messageID: String? = null
    ) {
        service?.sendPrivateMessage(content, recipientPeerID, recipientNickname, messageID)
        listener?.onSent(messageID, recipientPeerID)
    }

    fun establish(peerID: String) {
        service?.initiateNoiseHandshake(peerID)
    }

    fun isEstablished(peerID: String): Boolean = service?.hasEstablishedSession(peerID) == true

    fun sendFileBroadcast(file: BitchatFilePacket) {
        service?.sendFileBroadcast(file)
    }

    fun sendFilePrivate(recipientPeerID: String, file: BitchatFilePacket) {
        service?.sendFilePrivate(recipientPeerID, file)
    }

    fun cancelFileTransfer(transferId: String): Boolean {
        return service?.cancelFileTransfer(transferId) == true
    }

    fun peerNicknames(): Map<String, String> = service?.getPeerNicknames().orEmpty()

    fun peerRssi(): Map<String, Int> = service?.getPeerRSSI().orEmpty()

    private fun buildDelegate(): BluetoothMeshDelegate {
        return object : BluetoothMeshDelegate {
            override fun didReceiveMessage(message: BitchatMessage) {
                listener?.onMessageReceived(message)
                listener?.onReceived(message)
            }

            override fun didUpdatePeerList(peers: List<String>) {
                val newPeers = peers.toSet()
                val added = newPeers - lastPeers
                val removed = lastPeers - newPeers
                if (added.isNotEmpty()) {
                    added.forEach { peerID ->
                        listener?.onFound(peerID)
                        listener?.onConnected(peerID)
                    }
                }
                if (removed.isNotEmpty()) {
                    removed.forEach { peerID ->
                        listener?.onLost(peerID)
                        listener?.onDisconnected(peerID)
                    }
                }
                lastPeers = newPeers
                listener?.onPeerListUpdated(peers)
            }

            override fun didEstablishSession(peerID: String) {
                listener?.onEstablished(peerID)
            }

            override fun didUpdatePeerRSSI(peerID: String, rssi: Int) {
                listener?.onRSSIUpdated(peerID, rssi)
            }

            override fun didReceiveChannelLeave(channel: String, fromPeer: String) {
                // Not exposed in the minimal listener API.
            }

            override fun didReceiveDeliveryAck(messageID: String, recipientPeerID: String) {
                listener?.onDeliveryAck(messageID, recipientPeerID)
            }

            override fun didReceiveReadReceipt(messageID: String, recipientPeerID: String) {
                listener?.onReadReceipt(messageID, recipientPeerID)
            }

            override fun didReceiveVerifyChallenge(peerID: String, payload: ByteArray, timestampMs: Long) {
                listener?.onVerifyChallenge(peerID, payload, timestampMs)
            }

            override fun didReceiveVerifyResponse(peerID: String, payload: ByteArray, timestampMs: Long) {
                listener?.onVerifyResponse(peerID, payload, timestampMs)
            }

            override fun didReceiveFileTransfer(peerID: String, fileName: String, fileSize: Long, mimeType: String, localPath: String) {
                listener?.onFileReceived(peerID, fileName, fileSize, mimeType, localPath)
            }

            override fun decryptChannelMessage(encryptedContent: ByteArray, channel: String): String? {
                return null
            }

            override fun getNickname(): String? {
                return nickname
            }

            override fun isFavorite(peerID: String): Boolean {
                return false
            }
        }
    }
}
